File: packing_interface/src/algorithm_templates/cpp/root/ffdh.cpp
#include "packing_lib.h"
using namespace packing;

class Packing {
public:
    std::vector<std::tuple<double, double, int, int>> solve(
        int binWidth,
        const std::vector<std::tuple<int, int, int>>& rectangles
    ) {
        auto items = sort_by_height(expand_items(rectangles));

        struct Level {
            int y;
            int height;
            int usedWidth;
        };

        std::vector<Level> levels;
        std::vector<std::tuple<double, double, int, int>> placements;
        int currentY = 0;

        for (const auto& item : items) {
            bool placed = false;

            for (auto& level : levels) {
                if (level.usedWidth + item.width > binWidth) {
                    continue;
                }

                placements.push_back({
                    static_cast<double>(level.usedWidth),
                    static_cast<double>(level.y),
                    item.width,
                    item.height,
                });
                level.usedWidth += item.width;
                placed = true;
                break;
            }

            if (placed) {
                continue;
            }

            levels.push_back({currentY, item.height, item.width});
            placements.push_back({0.0, static_cast<double>(currentY), item.width, item.height});
            currentY += item.height;
        }

        return placements;
    }
};

File: packing_interface/src/algorithm_templates/python/root/ffdh.py
import packing_lib
from typing import List, Tuple

class Packing:
     def solve(self, bin_width: int, rectangles: List[Tuple[int, int, int]]) -> List[Tuple[float, float, int, int]]:       
        items = packing_lib.expand_items(rectangles)
        items = packing_lib.sort_by_height(items)
        levels = []
        placements = []
        current_y = 0

        for item in items:
            width = item["width"]
            height = item["height"]
            placed = False

            for level in levels:
                if level["used_width"] + width <= bin_width:
                    placements.append((level["used_width"], level["y"], width, height))
                    level["used_width"] += width
                    placed = True
                    break

            if placed:
                continue

            levels.append({"height": height, "used_width": width, "y": current_y})
            placements.append((0, current_y, width, height))
            current_y += height
        return packing_lib.output_from_placements(bin_width, placements)
